### DISCUSSION TITLE
Trouble with extensions

### DISCUSSION THREAD

**Discussion by ShobhanSrivastava** [kinngh/shopify-node-mongodb-next-app#53] [General]
Recently I migrated my app to this template with the latest version that comes with managed installation. My app already had 2 extensions, script extension and a pixel extension. Since this app doesn't explicitly tells you how to create extensions, I copied the extension directory from your express-mongodb template, and created the two extensions. Now when I am doing `npm run update:config`, the configuration for the app does change but the app is using the older extensions. I am really struggling with extensions, any help or a video tutorial would be really helpful.

**Comment by kinngh** [comment]
While a video + text update on this is coming on Monday, here's how to get this done:

1. Separate out your `app/` and `extension/` folders. (Rename your current folder to `app/` and create a new folder called `extension`
2. Update the `npm run update:config` script in `app/package.json` to `node _developer/tomlWriter.js; cd ..;shopify app deploy`
3. In your root, run `npm init --y` to ensure you have a `package.json` file in your root.
4. Get in your `app/` folder and run `npm run update:config` and it should write a `shopify.app.toml` file to your `extension/` folder. We're not deploying anything yet, but this ensures your setup is correct.
5. Place your `extensions/` in `extension/`
6. Go to your `app/` and run `npm run update:config` to push with your extension.

For reference, here's what the folder structure should look like:

```
app/
extension/
package.json
```

and this is a more detailed version:

```
app/
app/package.json //<--- npm run update:config lives here

extension/
extension/extensions/ // <---- all extensions live inside of this folder
extension/package.json

package.json //<--- this needs to exist or CLI would throw an error
```

**Reply by kinngh** [reply]
Simplified:
![Screenshot 2024-03-09 at 12 23 17 PM](https://github.com/kinngh/shopify-nextjs-prisma-app/assets/773555/462479bd-360f-49cb-aed7-b8b1c85ab5a1)

Detailed:
![Screenshot 2024-03-09 at 12 23 11 PM](https://github.com/kinngh/shopify-nextjs-prisma-app/assets/773555/2af3463f-fe9f-4c88-841c-9f15bbf72474)

**Comment by ShobhanSrivastava** [comment]
Did exactly what you mentioned. My folder structure looks like: 

<img width="276" alt="Screenshot 2024-03-09 at 2 20 45 PM" src="https://github.com/kinngh/shopify-nextjs-prisma-app/assets/76567059/306b6018-454a-4ade-b17f-96fa03a6ec23">

Running `npm run update:config` seems to have written 3 `shopify.app.toml` files. One each in the `root/`, `app/` and `extension/`. [Didn't release a new version at this point]

Now added `extensions/` folder inside the `extension/` 

<img width="262" alt="Screenshot 2024-03-09 at 2 23 19 PM" src="https://github.com/kinngh/shopify-nextjs-prisma-app/assets/76567059/aeec82d4-c0b2-493c-9fcc-68f7ed45f284">

From `app/` ran `npm run update:config` but it is not able to identify any extensions

<img width="227" alt="Screenshot 2024-03-09 at 2 24 47 PM" src="https://github.com/kinngh/shopify-nextjs-prisma-app/assets/76567059/ad70680a-25f5-4f89-9b80-c4bc3f4c079c">

**Reply by kinngh** [reply]
To confirm you did update the script to `"update:config": "node _developer/tomlWriter.js; cd ..;shopify app deploy"`, right?

**Reply by ShobhanSrivastava** [reply]
`"update:config": "node _developer/tomlWriter.js; cd ..; shopify app deploy",`

**Reply by kinngh** [reply]
Ahh I see what the problem is - `cd ..` is not honored by `npm`, but is honored by `bun`. There's two options here:

1. Switch to `bun` and run `bun run update:config` 
2. From your root, run `shopify app deploy` (if you do this, update script in `app/` to only write to toml and then deploy from root)

Essentially what's happening is we've told CLI that extension lives in `extension/extensions/**` and that needs to be consistent across the board or the CLI is going to have a rough time keeping everything in sync. The `toml` file technically is only correct in the root, so when you run `shopify app deploy` from root, it's going to work as expected. But since we update `toml` from our `tomlWriter`, it needs to be inside of the `app/` to be scoped properly and since `npm` doesn't honor changing current working directory, it's going to fail and not detect the extension.

![npm v bun](https://github.com/kinngh/shopify-nextjs-prisma-app/assets/773555/8781d757-92b3-4f26-9aff-79b200920365)

**Reply by ShobhanSrivastava** [reply]
Thanks man it worked. Cant tell you how helpful your videos and your prompt responses have been in my internship. This is my first time creating a Shopify app and you are a life saver!!

**Reply by kinngh** [reply]
Cheers mate!

**Reply by vedant14** [reply]
This helped me as well! Thanks :)

**Reply by kinngh** [reply]
I've added written instructions on how to add extensions - would be great to confirm if it's understandable: https://github.com/kinngh/shopify-nextjs-prisma-app/blob/main/docs/EXTENSIONS.md

**Reply by ShobhanSrivastava** [reply]
Ya it is

**Reply by AminRafaey** [reply]
Hi @kinngh, thank you for sharing details it worked locally but if we deploy over Vercel. Vercel does need next.js version inside root package.json file and there will not any So should we change the root directory inside Vercel to app/?

**Reply by kinngh** [reply]
@AminRafaey correct - you'll have to update your Vercel config by pointing it to the `app/` now instead of the root^

**Reply by AminRafaey** [reply]
@kinngh Changing the root directory over Vercel worked like a charm, but still, I can't see extension on my live shopify store. What specifically do we need to do to make it work over live with vercel?

Here is my vercel.json looks like:
`{
  "builds": [
    {
      "src": "app/package.json",
      "use": "@vercel/next",
      "config": {
        "distDir": "app/.next"
      }
    }
  ],
  "routes": [
    {
      "src": "/(.*)",
      "dest": "app/$1"
    }
  ]
}
`

**Reply by kinngh** [reply]
You do not deploy to Vercel for extensions. Just run "bun run update:config" with credentials from your live app and it pushes from your local to prod since extensions are hosted by Shopify, not by the developer.

**Comment by ShobhanSrivastava** [comment]
Hey @kinngh, is there a way to inspect the code inside the extensions after it goes to Shopify's CDN?

**Reply by kinngh** [reply]
Why would you want to do that?

**Reply by ShobhanSrivastava** [reply]
I've run into an issue. I have 2 extensions, a web pixel extension and a theme app extension and I updated both of them using the steps that you mentioned above. But strangely both my extensions have the last updated dates in the past.

<img width="837" alt="Screenshot 2024-03-11 at 3 23 09 PM" src="https://github.com/kinngh/shopify-nextjs-prisma-app/assets/76567059/aa1d0352-5786-45d4-9fc0-01ecfd39eb87">
<img width="831" alt="Screenshot 2024-03-11 at 3 24 08 PM" src="https://github.com/kinngh/shopify-nextjs-prisma-app/assets/76567059/206bcbd0-222e-4e8e-a5b5-a45384ea1980">

And also the theme app extension surprisingly doesn't work. 

Now to make it more clear, I have 2 apps, Prod and a Local app. On the local app (For testing purposes) the last update dates are in the past as well but everything seems to be working fine. But its the PROD app on which the theme app extension is not working. How to make sure that everytime the app is updated the extensions are also updated whether any change has been made or not?

**Reply by ShobhanSrivastava** [reply]
Could it be a caching issue on Shopify's end?

**Reply by kinngh** [reply]
Could be a caching issue, but since you're technically running `shopify app deploy` and it's not throwing any errors, it should be fine and sometimes these caching issues happen. Try in a fresh app to see if everything's going through?

**Reply by ShobhanSrivastava** [reply]
Yes in a fresh app the changes are going through. Is there a way to make sure that the changes always go through?

**Reply by kinngh** [reply]
I don't see problems on my end on any of our apps - try getting through partner support? One thing you can do is when you're making requests or are in specific stores, console.log() a version string so you know what version is being used, or you can always add something to `window` in your theme extension so you can run `window?.app_name?.version` and get a version string off it and confirm if it's cache or it's not going through

**Reply by ShobhanSrivastava** [reply]
So in order to solve that caching issue, I added a package.json with the following data in each of my extensions directory. 

`{
  "name": "pixel-extension",
  "version": "1.0.0"
}`

Now on every change I am making sure to update the version inside my `package.json`. After this, running `npm run deploy` does change the extensions. 

Can you please review and let me know if its the right approach? 

Now one more issue, the extension isn't updated for the users that had my app previously installed. The extension changes seem to show up only for the users installing my app thereafter.

**Reply by kinngh** [reply]
- Tbh if it works, it works ™ 
- Are you sure there's a new release or have you modified something? This is extremely odd. Can you DM me on [X](https://x.com/kinngh)? I want to see this

**Reply by ShobhanSrivastava** [reply]
So I tried the same with my test app, but it did not work this time.
